### PR TITLE
Add file_list support

### PR DESCRIPTION
**Pull request by fizzi01**
This PR adds support for copying and pasting file lists (i.e., lists of file paths) to and from the clipboard. It simply utilizes the already implemented functionality in `arboard`.

**Testing:**

* Added platform-specific helpers for reading and writing file lists to the clipboard in `tests/clipboard.py` for macOS (using AppKit), Windows (using PowerShell), and Linux (using xclip and URI encoding).
* Added `pyobjc` as a development dependency for macOS to enable native clipboard access for file lists. I couldn't get osascript to do this.

~~PS: I've add an osascript fallback but it just doesn't work with multiple files, it returns the last one path~~

### CODE REVIEW

**Comment by Klavionik** [comment]
Hey, thanks for the PR! I'm quite busy with personal stuff lately, but I'll try and get to it as soon as I can.

File: python/copykitten/__init__.py [hunk 1]
```diff
@@ -8,8 +8,11 @@
 from ._copykitten import copy_wait as _copy_wait
 from ._copykitten import paste as _paste
 from ._copykitten import paste_image as _paste_image
+from ._copykitten import paste_file_list as _paste_file_list
+from ._copykitten import copy_file_list as _copy_file_list
+from ._copykitten import copy_file_list_wait as _copy_file_list_wait
 
-__all__ = ["copy", "paste", "clear", "copy_image", "paste_image", "CopykittenError"]
+__all__ = ["copy", "paste", "clear", "copy_image", "paste_image", "paste_file_list", "copy_file_list", "CopykittenError"]
```
Comment (Klavionik) [inline python/copykitten/__init__.py]: Just for the sake of consistency, may I ask you to put `copy_file_list` before `paste_file_list`? The same goes for `__init__.py`.

File: python/copykitten/_copykitten.pyi [hunk 2]
```diff
@@ -1,10 +1,12 @@
...
 
-
 def copy(content: str) -> None: ...
 def copy_wait(content: str) -> None: ...
 def paste() -> str: ...
-def clear() -> None:...
+def clear() -> None: ...
 def copy_image(content: bytes, width: int, height: int) -> None: ...
 def copy_image_wait(content: bytes, width: int, height: int) -> None: ...
+def copy_file_list(file_list: list[str]) -> None: ...
```
Comment (Klavionik) [inline python/copykitten/_copykitten.pyi]: Again, could you put these two signatures just before `paste_file_list`?

That way, we get a cluster of text-related functions, then image-related, then file-list-related.

File: src/lib.rs [hunk 3]
```diff
@@ -114,6 +114,35 @@ fn copy_image_wait(content: Cow<[u8]>, width: usize, height: usize) -> PyResult<
...
+#[cfg(not(target_os = "linux"))]
+#[pyfunction]
+fn copy_file_list_wait(file_list: Vec<String>) -> PyResult<()> {
+    copy_file_list(file_list)
+}
+
+#[cfg(target_os = "linux")]
+#[pyfunction]
+fn copy_file_list_wait(file_list: Vec<String>) -> PyResult<()> {
+    let paths: Vec<std::path::PathBuf> = file_list.into_iter().map(std::path::PathBuf::from).collect();
```
Comment (Klavionik) [inline src/lib.rs]: I feel like there must be a better way.

What do you think if we make the API employ [Postel's law](https://en.wikipedia.org/wiki/Robustness_principle)? We would accept `list[str | os.PathLike]` so that the API works with `pathlib.Path` instances as well, and we would return `list[pathlib.Path]` to make it more type-safe and self-documenting (at this point I'm also questioning myself, how common and idiomatic this is).

`pyo3` docs says that it can [implicitly convert](https://pyo3.rs/v0.20.0/conversions/tables) between Python types and `std::path::PathBuf`, that would be of great help here.

Feel free to share your thoughts about the proposed approach vs. operating pure strings.

**Comment by Klavionik** [comment]
I have finally found some time to at least skim over the code. 😥 Didn't make it to the tests though.

May I ask you to elaborate on the use case for this? What's the benefit for the user? Like, can we now programmatically copy file URIs to the clipboard so that the user can paste them later into a directory or is it something else?

**Comment by fizzi01** [comment]
> I have finally found some time to at least skim over the code. 😥 Didn't make it to the tests though.
> 
> May I ask you to elaborate on the use case for this? What's the benefit for the user? Like, can we now programmatically copy file URIs to the clipboard so that the user can paste them later into a directory or is it something else?

thx for your time. If someone copies multiple files, you can't access the absolute path of them to do some elaborations. For example, for a file-sharing use case, this type of information is needed to access the correct files. Actually i don't know if pasting URIs, let the os automatically handle pasting files later, i need to check that. As for now reading the copied paths is more useful.

File: src/lib.rs (hunk 3 above)
Comment (fizzi01) [reply src/lib.rs]: i definitely agree, there is no need to force using just strings

**Comment by fizzi01** [comment]
@Klavionik The error on macOS tests should be fixed now, i thought that pyobjc wasn't available for python 3.8, but i was wrong.

File: src/lib.rs [hunk 4]
```diff
@@ -114,6 +115,33 @@ fn copy_image_wait(content: Cow<[u8]>, width: usize, height: usize) -> PyResult<
     copy_image(content, width, height)
 }
 
+#[pyfunction]
+fn copy_file_list(file_list: Vec<std::path::PathBuf>) -> PyResult<()> {
+    let mut cb = get_clipboard()?;
+    let file_list: Vec<&std::path::Path> = file_list.iter().map(|p| p.as_path()).collect();
```
Comment (Klavionik) [inline src/lib.rs:121]: Luckily for us, `PathBuf` implements `AsRef<Path>`, so you don't need to do the mapping manually.

File: tests/clipboard.py [hunk 5]
```diff
@@ -74,6 +80,36 @@ def write_image_macos(img: Image.Image) -> None:
         subprocess.run(cmd, check=True)
 
 
+def _require_appkit():
+    try:
+        from AppKit import NSPasteboard, NSFilenamesPboardType
+
+        return NSPasteboard, NSFilenamesPboardType
+    except (ImportError, ModuleNotFoundError):
```
Comment (Klavionik) [inline tests/clipboard.py]: `ModuleNotFoundError` is a subclass of `ImportError`, so you don't have to catch both here.

A more important question is, why do we want to skip some tests if a dev requirement is not satisfied? I feel like we should not, but maybe I'm missing something.

File: tests/clipboard.py [hunk 6]
```diff
@@ -131,6 +167,61 @@ def write_image_win(img: Image.Image) -> None:
...
+    )
+    subprocess.run(("powershell.exe", "-NoProfile", "-Command", script), check=True)
+
+
+def read_file_list_linux() -> List[str]:
+    result = subprocess.check_output(
+        ("xclip", "-sel", "clipboard", "-o", "-target", "text/uri-list")
+    )
+    paths = []
+    for line in result.decode().splitlines():
```
Comment (Klavionik) [inline tests/clipboard.py:198]: Could you fill me in on what's exactly happening here? Why perform a `line.strip()`, why all the checks? Just trying to figure out what kind of corner cases we are solving here.

File: tests/clipboard.py [hunk 7]
```diff
@@ -131,6 +167,61 @@ def write_image_win(img: Image.Image) -> None:
...
+            paths.append(unquote(line[7:]))
+        else:
+            paths.append(line)
+    return paths
+
+
+def write_file_list_linux(file_list: List[str]) -> None:
+    from urllib.parse import quote
+
+    uri_list = "\n".join("file://" + quote(p, safe="/") for p in file_list)
```
Comment (Klavionik) [inline tests/clipboard.py:214]: I feel like `Path.as_uri` might be helpful here.

File: tests/conftest.py [hunk 8]
```diff
@@ -42,6 +44,16 @@ def write_clipboard_image(clipboard) -> WriteClipboardImage:
     return clipboard.write_image
 
 
+@pytest.fixture(scope="session")
+def paste_file_list(clipboard) -> WriteClipboardFileList:
+    return clipboard.write_file_list
+
+
+@pytest.fixture(scope="session")
+def read_clipboard_file_list(clipboard) -> ReadClipboardFileList:
```
Comment (Klavionik) [inline tests/conftest.py:53]: Could you swap the order of the added fixtures, please? First read, then paste.

File: tests/test_lib.py [hunk 9]
```diff
@@ -95,7 +98,44 @@ def test_copy_image(test_image: Image.Image, read_clipboard_image: ReadClipboard
...
+
+    actual = read_clipboard_file_list()
+
+    assert actual == file_list
+
+
+def test_copy_file_list_pathlike(
+    tmp_path,
+    read_clipboard_file_list: ReadClipboardFileList,
+    paste_file_list: WriteClipboardFileList,
```
Comment (Klavionik) [inline tests/test_lib.py]: Unused.

File: tests/test_lib.py [hunk 10]
```diff
@@ -164,6 +226,29 @@ def test_copy_image_detach(test_image: Image.Image, read_clipboard_image: ReadCl
     assert pasted_image.tobytes() == test_image.tobytes()
 
 
+@pytest.mark.skipif(sys.platform != "linux", reason="Detach is supported only on Linux")
+def test_copy_file_list_detach(
+    tmp_path,
+    read_clipboard_file_list: ReadClipboardFileList,
+    paste_file_list: WriteClipboardFileList,
```
Comment (Klavionik) [inline tests/test_lib.py]: Unused.

File: tests/clipboard.py [hunk 11]
```diff
@@ -140,7 +231,9 @@ def write_linux(content: str) -> None:
 
 
 def read_image_linux() -> Image.Image:
-    data = subprocess.check_output(("xclip", "-sel", "clipboard", "-o", "-target", "image/png"))
+    data = subprocess.check_output(
```
Comment (Klavionik) [inline tests/clipboard.py]: Looks like an accidental change.

File: tests/clipboard.py [hunk 12]
```diff
@@ -48,7 +52,9 @@ def write_macos(content: str) -> None:
 
 
 def read_image_macos() -> Image.Image:
-    data = subprocess.check_output(("osascript", "-e", "get the clipboard as «class PNGf»"))
+    data = subprocess.check_output(
```
Comment (Klavionik) [inline tests/clipboard.py]: Looks like an accidental change.

File: tests/clipboard.py [hunk 13]
```diff
@@ -37,6 +39,8 @@ class Clipboard:
     write = Resolver[WriteClipboard]()
     read_image = Resolver[ReadClipboardImage]()
     write_image = Resolver[WriteClipboardImage]()
+    write_file_list = Resolver[WriteClipboardFileList]()
```
Comment (Klavionik) [inline tests/clipboard.py]: First read, then write.

File: tests/clipboard.py [hunk 14]
```diff
@@ -3,14 +3,16 @@
...
-from typing import Callable, Generic, TypeVar, cast
+from typing import Callable, Generic, TypeVar, cast, List
 
 from PIL import Image
 
 ReadClipboard = Callable[[], str]
 WriteClipboard = Callable[[str], None]
 ReadClipboardImage = Callable[[], Image.Image]
 WriteClipboardImage = Callable[[Image.Image], None]
+WriteClipboardFileList = Callable[[List[str]], None]
```
Comment (Klavionik) [inline tests/clipboard.py:15]: First read, then write.

File: src/lib.rs [hunk 15]
```diff
@@ -4,10 +4,11 @@ extern crate core;
 use arboard::SetExtLinux;
 #[cfg(target_os = "linux")]
 use daemonize::{Daemonize, Outcome};
+#[cfg(target_os = "linux")]
+use std::fs::File;
```
Comment (Klavionik) [inline src/lib.rs]: Nice catch!

File: tests/test_lib.py [hunk 16]
```diff
@@ -95,7 +98,44 @@ def test_copy_image(test_image: Image.Image, read_clipboard_image: ReadClipboard
     assert test_image_bytes == pasted_image.tobytes()
 
 
-def test_paste_image(test_image: Image.Image, write_clipboard_image: WriteClipboardImage):
+def test_copy_file_list(
+    tmp_path,
```
Comment (Klavionik) [inline tests/test_lib.py]: Missing a type hint.

File: tests/clipboard.py [hunk 17]
```diff
@@ -3,14 +3,16 @@
...
+ReadClipboardFileList = Callable[[], List[str]]
```
Comment (Klavionik) [inline tests/clipboard.py]: Maybe return a list of `Paths`? Would it be more in line with the API under test and therefore more convenient to use?

File: python/copykitten/__init__.py (hunk 1 above)
Comment (Klavionik) [reply python/copykitten/__init__.py]: Sorry, I didn't make myself clear last time. Should have said "the same goes for the function definitions below in this file".

**Review by Klavionik — CHANGES_REQUESTED** [review]

**Comment by Klavionik** [comment]
@fizzi01 Finally had time to properly review the code. I like where we are going!

 >Actually i don't know if pasting URIs, let the os automatically handle pasting files later, i need to check that.

Have you managed to do this?

**Comment by Klavionik** [comment]
I have introduced some changes to `main`, so please rebase your branch (and expect the lint workflow to be somewhat unhappy as it has to do actual work now :) ).

**Comment by Klavionik** [comment]
@fizzi01 Hey, it's been quite a while. Any chance you'll get back on track soon?

**Comment by fizzi01** [comment]
> @fizzi01 Hey, it's been quite a while. Any chance you'll get back on track soon?

Yep sorry, starting from next week. I've been quite busy

**Comment by fizzi01** [comment]
> @fizzi01 Finally had time to properly review the code. I like where we are going!
> 
> > Actually i don't know if pasting URIs, let the os automatically handle pasting files later, i need to check that.
> 
> Have you managed to do this?

I confirm that it does work as intended. The os does it automatically! (tried only on macOs actually)

File: python/copykitten/__init__.py [hunk 18]
```diff
@@ -83,3 +97,28 @@ def paste_image() -> Tuple[bytes, int, int]:
     :return: A 3-tuple of raw RGBA pixels, width, and height.
     """
     return _paste_image()
+
+
+def copy_file_list(file_list: List[Union[str, os.PathLike]], *, detach: bool = False) -> None:
```
Comment (Klavionik) [inline python/copykitten/__init__.py]: Type argument for `os.PathLike` is missing. I suggest we use `os.PathLike[str]`.

File: python/copykitten/__init__.py [hunk 19]
```diff
@@ -83,3 +97,28 @@ def paste_image() -> Tuple[bytes, int, int]:
...
+    """
+    Copies a list of file paths into the clipboard.
+
+    :param file_list: A list of file paths to copy. Accepts strings and os.PathLike objects.
```
Comment (Klavionik) [inline python/copykitten/__init__.py]: The second sentence is redundant given the presence of the type hints.

File: python/copykitten/__init__.py [hunk 20]
```diff
@@ -83,3 +97,28 @@ def paste_image() -> Tuple[bytes, int, int]:
...
+    else:
+        _copy_file_list(file_list)
+
+
+def paste_file_list() -> List[pathlib.Path]:
+    """
+    Returns a list of file paths from the clipboard.
+
+    :raises CopykittenError: Raised if there's no file list in the clipboard.
+    :return: A list of file paths as pathlib.Path objects.
```
Comment (Klavionik) [inline python/copykitten/__init__.py]: `A list of file paths` seems to be enough to me, no need to repeat the type hint.

File: python/copykitten/_copykitten.pyi [hunk 21]
```diff
@@ -1,10 +1,15 @@
...
 def copy_wait(content: str) -> None: ...
 def paste() -> str: ...
-def clear() -> None:...
+def clear() -> None: ...
 def copy_image(content: bytes, width: int, height: int) -> None: ...
 def copy_image_wait(content: bytes, width: int, height: int) -> None: ...
 def paste_image() -> tuple[bytes, int, int]: ...
+def copy_file_list(file_list: list[str | os.PathLike]) -> None: ...
+def copy_file_list_wait(file_list: list[str | os.PathLike]) -> None: ...
+def paste_file_list() -> list[pathlib.Path]: ...
```
Comment (Klavionik) [inline python/copykitten/_copykitten.pyi]: `os.PathLike` -> `os.PathLike[str]` (see other comment)

File: tests/clipboard.py [hunk 22]
```diff
@@ -74,6 +78,36 @@ def write_image_macos(img: Image.Image) -> None:
...
 
+def _require_appkit():
+    try:
+        from AppKit import NSFilenamesPboardType, NSPasteboard
+
+        return NSPasteboard, NSFilenamesPboardType
+    except ImportError:
+        import pytest
+
+        pytest.skip("AppKit not available (requires pyobjc)")
```
Comment (Klavionik) [inline tests/clipboard.py]: We're still skipping tests if dev dependencies aren't satisfied. I'd rather tests fail in that case. And then we don't really need this helper.

File: tests/clipboard.py (hunk 6 above)
Comment (Klavionik) [reply tests/clipboard.py]: Still no answer to this question and no changes in the code.

File: tests/clipboard.py (hunk 7 above)
Comment (Klavionik) [reply tests/clipboard.py]: Still pending action.

File: tests/conftest.py (hunk 8 above)
Comment (Klavionik) [reply tests/conftest.py]: Still pending action.

File: tests/test_lib.py [hunk 23]
```diff
@@ -95,6 +98,24 @@ def test_copy_image(test_image: Image.Image, read_clipboard_image: ReadClipboard
     assert test_image_bytes == pasted_image.tobytes()
 
 
+def test_copy_file_list(
```
Comment (Klavionik) [inline tests/test_lib.py:101]: It would be cool to have a detach=True test case covered.

File: src/lib.rs (hunk 4 above)
Comment (Klavionik) [reply src/lib.rs]: Still pending action.

**Review by Klavionik — COMMENTED** [review]
Hey, sorry it took so long. There are still a couple of things to fix, plus some important comments from the previous review were not addressed. 

As this PR is really overdue, I think I'll go ahead and merge it soon, and then work on the required changes myself since I have some extra time.

File: tests/test_lib.py (hunk 9 above)
Comment (Klavionik) [reply tests/test_lib.py]: The test case was just fine. :) The comment was in regard to the `paste_file_list` parameter being unused.

File: tests/test_lib.py (hunk 10 above)
Comment (Klavionik) [reply tests/test_lib.py]: The test case was just fine. :) The comment was in regard to the `paste_file_list` parameter being unused. [2]

File: tests/clipboard.py (hunk 6 above)
Comment (fizzi01) [reply tests/clipboard.py]: It was just a quick test for xclip. Forgot to improve the code :(

**Comment by Klavionik** [comment]
Looks super solid now! The only thing left is to fix the linter error and we're good to go. Could you do that please?

**Comment by fizzi01** [comment]
> Looks super solid now! The only thing left is to fix the linter error and we're good to go. Could you do that please?

Done!
